win32ss/user/ntuser/winpos.c

```diff
@@ -3709,35 +3708,43 @@ NtUserSetWindowPlacement(HWND hWnd,
    TRACE("Enter NtUserSetWindowPlacement\n");
    UserEnterExclusive();
 
-   if (!(Wnd = UserGetWindowObject(hWnd)) ||
-        UserIsDesktopWindow(Wnd) || UserIsMessageWindow(Wnd))
-   {
-      goto Exit; // Return FALSE
-   }
+    _SEH2_TRY
+    {
+        ProbeForRead(lpwndpl, sizeof(WINDOWPLACEMENT), 1);
+        Safepl = *lpwndpl;
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        SetLastNtError(_SEH2_GetExceptionCode());
+        _SEH2_YIELD(goto Exit); // Return FALSE
+    }
+    _SEH2_END
 
-   _SEH2_TRY
-   {
-      ProbeForRead(lpwndpl, sizeof(WINDOWPLACEMENT), 1);
-      RtlCopyMemory(&Safepl, lpwndpl, sizeof(WINDOWPLACEMENT));
-   }
-   _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
-   {
-      SetLastNtError(_SEH2_GetExceptionCode());
-      _SEH2_YIELD(goto Exit); // Return FALSE
-   }
-   _SEH2_END
+    /* Backwards-compatibility: Win 3.x doesn't check the length */
+    if (LOWORD(gptiCurrent->dwExpWinVer) < WINVER_WINNT4)
+        Safepl.length = sizeof(WINDOWPLACEMENT);
 
-   if(Safepl.length != sizeof(WINDOWPLACEMENT))
-   {
-      goto Exit; // Return FALSE
-   }
+    if (Safepl.length == sizeof(WINDOWPLACEMENT))
+    {
+        ERR("0x%lX, 0x%lX\n", gptiCurrent->dwExpWinVer, WINVER_WINNT4);
+        Flags = PLACE_MAX | PLACE_RECT;
+        if (Safepl.flags & WPF_SETMINPOSITION)
+            Flags |= PLACE_MIN;
+    }
+    else
+    {
+        EngSetLastError(ERROR_INVALID_PARAMETER);
+        goto Exit;
+    }
```

Comment:
```suggestion
    if (Safepl.length != sizeof(WINDOWPLACEMENT))
    {
        EngSetLastError(ERROR_INVALID_PARAMETER);
        goto Exit;
    }

    Flags = PLACE_MAX | PLACE_RECT;
    if (Safepl.flags & WPF_SETMINPOSITION)
        Flags |= PLACE_MIN;
```

Also, I don't think the `ERR("0x%lX, 0x%lX\n", gptiCurrent->dwExpWinVer, WINVER_WINNT4);` is needed anymore (it was a debug thing for checking whether dwExpWinVer was being set), so I removed it in the suggestion.